File: searching/binary_search/binary_search.py
def binary_search_iterative(elements: list[int], key: int) -> int:
    if not elements:
        return -1
    left: int = 0
    right: int = len(elements)

    while left <= right:
        mid: int = left + (right - left) // 2

        if elements[mid] == key:
            return mid

        if key > elements[mid]:
            left = mid + 1
        else:
            right = mid - 1

    return -1


if __name__ == '__main__':
    input: list[int] = [4, 10, 9, 0, 3, 11, 5, 8, 5, 3, 5, 7, 8, 12, 2]
    input.sort()
    key: int = 12

    print(f"{input = }")

    print(f"{key = }")
    print(f"{binary_search_iterative(input, key) = }")

    key: int = 0
    print(f"{key = }")
    print(f"{binary_search_iterative(input, key) = }")

    key: int = 8
    print(f"{key = }")
    print(f"{binary_search_iterative(input, key) = }")

File: searching/binary_search/BinarySearch.java
public class BinarySearch {
    public static void main(String[] args) {
        int[] input = { 0, 2, 3, 6, 7, 9 };
        int key = 9;
        print(input, "Input array:");
        System.out.println("key: " + key);
        System.out.println("binarySearch: " + binarySearch(input.clone(), key));

        key = 0;
        System.out.println("key: " + key);
        System.out.println("binarySearch: " + binarySearch(input.clone(), key));

        key = 10;
        System.out.println("key: " + key);
        System.out.println("binarySearch: " + binarySearch(input.clone(), key));

    }

    public static int binarySearch(int[] elements, int key) {
        int left = 0;
        int right = elements.length - 1;
        int middle;
        while (left <= right) {
            middle = left + (right - left) / 2;
            if (key == elements[middle])
                return middle;

            if (key < elements[middle]) {
                right = middle - 1;
            } else {
                left = middle + 1;
            }
        }
        return -1;
    }

    static void print(int[] array, String text) {
        System.out.print(text + " ");
        for (int num : array) {
            System.out.print(num + " ");
        }
        System.out.println();
    }

    static void print(int[] array) {
        for (int num : array) {
            System.out.print(num + " ");
        }
        System.out.println();
    }
}